### PR TITLE
remove some unnecessary cudaGetDevices

### PR DESCRIPTION
**Pull request by ngimel**
Removes some (not all) unnecessary cudaGetDevice's. It does not affect usual performance, except in may be some pathological cases, but it makes timings under nvprof slightly less distorted.

### CODE REVIEW

File: aten/src/THC/THCApply.cuh [hunk 1]
```diff
@@ -165,13 +165,15 @@ inline dim3 getApplyBlock() {
   return dim3(THC_APPLY_THREADS_PER_BLOCK);
 }
 
-inline bool getApplyGrid(THCState* state, uint64_t totalElements, dim3& grid) {
-  int curDevice = -1;
-  cudaGetDevice(&curDevice);
-  if (curDevice == -1) return false;
+inline bool getApplyGrid(THCState* state, uint64_t totalElements, dim3& grid, int curDevice = -1) {
```
Comment (apaszke) [inline aten/src/THC/THCApply.cuh]: Does this argument really needs to be optional? It's always specified

File: aten/src/THC/THCApply.cuh [hunk 2]
```diff
@@ -236,7 +240,7 @@ bool THC_pointwiseApply1(THCState* state,
   kernelPointwiseApply1<Op,                                             \
                         typename TensorUtils<TensorTypeA>::DataType,    \
                         TYPE, A>                                        \
-    <<<grid, block, 0, THCState_getCurrentStream(state)>>>(             \
+    <<<grid, block, 0, THCState_getCurrentStreamOnDevice(state, curDevice)>>>(             \
```
Comment (apaszke) [inline aten/src/THC/THCApply.cuh]: You're never checking that `curDevice` is valid (it wasn't an error)

**Review by apaszke — COMMENTED** [review]
A random idea that would let us eliminate a bunch of those calls without threading the device id everywhere manually: what if we had a `thread_local` value where we would stash the current device for as long as we *know* that we'll be entirely in code known to us? We could manage that with a RAII style structs, to make sure the cache would get invalidated upon exit from any of those functions.

File: aten/src/THC/THCApply.cuh (hunk 1 above)
Comment (ngimel) [reply aten/src/THC/THCApply.cuh]: Yes, there are a few places where getApplyGrid is called that I did not patch (FusedRNNCell and may be something else)

File: aten/src/THC/THCApply.cuh (hunk 2 above)
Comment (ngimel) [reply aten/src/THC/THCApply.cuh]: You will never get here if there was an error, getApplyGrid will return an error.

Comment (apaszke) [reply aten/src/THC/THCApply.cuh]: Hmm ok. Making error checking part of `getApplyGrid`'s contract is quite weird 😕 Can we add an `AT_CHECK` there? It's pretty much free.

**Comment by ngimel** [comment]
@apaszke having a thread-local cached device id would be great. That's exactly what cudaGetDevice should be doing, it does not have to be such a pain in the neck, but well, it is.

**Comment by ngimel** [comment]
c2 CI failures seem unrelated?

**Review by soumith — APPROVED** [review]

**Pull request merged**